data/abi/abi.go

```diff
@@ -0,0 +1,1164 @@
+}
+
+// MakeUintType makes `Uint` ABI type by taking a type size argument.
+// The range of type size is [8, 512] and type size % 8 == 0.
+func MakeUintType(typeSize uint16) (Type, error) {
+	if typeSize%8 != 0 || typeSize < 8 || typeSize > 512 {
+		return Type{}, fmt.Errorf("unsupported uint type size: %d", typeSize)
+	}
+	return Type{
+		enumIndex: Uint,
+		size:      typeSize,
+	}, nil
+}
+
+// MakeByteType makes `Byte` ABI type.
+func MakeByteType() Type {
+	return Type{
+		enumIndex: Byte,
+	}
+}
+
+// MakeUfixedType makes `UFixed` ABI type by taking type size and type precision as arguments.
+// The range of type size is [8, 512] and type size % 8 == 0.
+// The range of type precision is [1, 160].
+func MakeUfixedType(typeSize uint16, typePrecision uint16) (Type, error) {
+	if typeSize%8 != 0 || typeSize < 8 || typeSize > 512 {
+		return Type{}, fmt.Errorf("unsupported ufixed type size: %d", typeSize)
+	}
+	if typePrecision > 160 || typePrecision < 1 {
+		return Type{}, fmt.Errorf("unsupported ufixed type precision: %d", typePrecision)
+	}
+	return Type{
+		enumIndex: Ufixed,
+		size:      typeSize,
+		precision: typePrecision,
+	}, nil
+}
+
+// MakeBoolType makes `Bool` ABI type.
+func MakeBoolType() Type {
+	return Type{
+		enumIndex: Bool,
+	}
+}
+
+// MakeStaticArrayType makes static length array ABI type by taking
+// array element type and array length as arguments.
+func MakeStaticArrayType(argumentType Type, arrayLength uint16) Type {
+	return Type{
+		enumIndex:    ArrayStatic,
+		childTypes:   []Type{argumentType},
+		staticLength: arrayLength,
+	}
+}
+
+// MakeAddressType makes `Address` ABI type.
+func MakeAddressType() Type {
+	return Type{
+		enumIndex: Address,
+	}
+}
+
+// MakeDynamicArrayType makes dynamic length array by taking array element type as argument.
+func MakeDynamicArrayType(argumentType Type) Type {
+	return Type{
+		enumIndex:  ArrayDynamic,
+		childTypes: []Type{argumentType},
+	}
+}
+
+// MakeStringType makes `String` ABI type.
+func MakeStringType() Type {
+	return Type{
+		enumIndex: String,
+	}
+}
+
+// MakeTupleType makes tuple ABI type by taking an array of tuple element types as argument.
+func MakeTupleType(argumentTypes []Type) (Type, error) {
+	if len(argumentTypes) >= (1 << 16) {
+		return Type{}, fmt.Errorf("tuple type child type number >= 2^16 error")
+	}
+	return Type{
+		enumIndex:    Tuple,
+		childTypes:   argumentTypes,
+		staticLength: uint16(len(argumentTypes)),
+	}, nil
+}
+
+// Equal method decides the equality of two types: t == t0.
+func (t Type) Equal(t0 Type) bool {
+	if t.enumIndex != t0.enumIndex {
+		return false
+	} else if t.precision != t0.precision || t.size != t0.size {
+		return false
+	} else if t.staticLength != t0.staticLength {
+		return false
+	} else {
+		if len(t.childTypes) != len(t0.childTypes) {
+			return false
+		}
+		for i := 0; i < len(t.childTypes); i++ {
+			if !t.childTypes[i].Equal(t0.childTypes[i]) {
+				return false
+			}
+		}
+	}
+	return true
+}
```

Comment:
Thanks for the review, now this `Equal` tests the struct fields, which makes things more straightforward.